adding a blog about statefulset (#9776)
istio/istio.io: content/en/blog/2021/stateful-set-made-easier/index.md

---
title: "StatefulSets Made Easier With Istio 1.10"
description: Learn how to easily deploy stateful sets with Istio 1.10.
publishdate: 2021-05-18
attribution: "Lin Sun (Solo.io), Christian Posta (Solo.io), John Howard (Google), Zhonghu Xu (Huawei)"
keywords: [statefulset,Istio,networking,localhost,loopback,eth0]
---

Kubernetes [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) are commonly used to manage stateful applications. In addition to managing the deployment and scaling of a set of Pods, StatefulSets provide guarantees about the ordering and uniqueness of these Pods. Common StatefulSets used with Istio are ZooKeeper, Cassandra, Elasticsearch, Redis, Apache NiFi, and more. The Istio community has been making gradual progress towards supporting zero-configuration for StatefulSets, from automatic mTLS to eliminating any ServiceEntry or Istio resources, to the most recent [pod networking change in Istio 1.10](/blog/2021/upcoming-networking-changes/).

## Challenges With StatefulSet

> 

Author: updated

> Many [issues](https://github.com/istio/istio/issues/10659) have been reported from the Istio community around StatefulSets. What is unique about StatefulSet and service mesh? The StatefulSet pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. Pods within a StatefulSet listen on the pod IP only instead of `0.0.0.0` for communication within the StatefulSet pods.

Author: This is true for stateful sets in general.  also mentioned in @howardjohn 's blog. I think it is related to how stateful sets pods are created from the same spec and didn't want to listen on loopback interface for that inter-pod communication.

> Many [issues](https://github.com/istio/istio/issues/10659) have been reported from the Istio community around StatefulSets. What is unique about StatefulSet and service mesh? The StatefulSet pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. Pods within a StatefulSet listen on the pod IP only instead of `0.0.0.0` for communication within the StatefulSet pods.

Reviewer: This is not accurate, any application, statefulset or otherwise, can bind to any interface it chooses. It just happens to be common among applications that happen to use statefulsets

> Many [issues](https://github.com/istio/istio/issues/10659) have been reported from the Istio community around StatefulSets. What is unique about StatefulSet and service mesh? The StatefulSet pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. Pods within a StatefulSet listen on the pod IP only instead of `0.0.0.0` for communication within the StatefulSet pods.

Author: @craigbox @howardjohn anything you propose to change here?  I didn't cover all pods/apps as John covered in his other blog.  I think what we have around stateful sets is accurate in this blog.

> Many [issues](https://github.com/istio/istio/issues/10659) have been reported from the Istio community around StatefulSets. What is unique about StatefulSet and service mesh? The StatefulSet pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. Pods within a StatefulSet listen on the pod IP only instead of `0.0.0.0` for communication within the StatefulSet pods.

Reviewer: `Pods within a StatefulSet listen on the pod IP only instead of `0.0.0.0` for communication within the StatefulSet pods.` is not accurate. It would be accurate to say `It is common for pods within a StatefulSet listen on the pod IP only, instead of 0.0.0.0, for communication within the StatefulSet pods`, but it certainly doesn't apply to all statefulsets.

Statefulsets are actually just normal pods with a few fields set BTW, you can do everything with a statefulset with Pod

> Many [issues](https://github.com/istio/istio/issues/10659) have been reported from the Istio community around StatefulSets. What is unique about StatefulSet and service mesh? The StatefulSet pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. Pods within a StatefulSet listen on the pod IP only instead of `0.0.0.0` for communication within the StatefulSet pods.

Author: Yah, I like that, thanks for that clarification! you'll never know how users can configure a StatefulSet :)

Using ZooKeeper as an example, from the `zoo.cfg` file in my ZooKeeper pod, it is configured not to listen on all IPs for quorum communication within the StatefulSet pods.

{{< text plain >}}
quorumListenOnAllIPs=false

> {{< /text >}}

Author: Actually, it is not that simple. :) I think ZK doesn't recommend users to enable this in production.  This quorumListenOnAllIPs is an unsafe option per their doc, even tho our istio [doc](https://istio.io/latest/faq/applications/#zookeeper) recommended users to do it today in 1.9, which is bad.

> {{< /text >}}

Reviewer 2: note: someone should update the FAQ

> {{< /text >}}

Author: Yes, plan to update it after this is merged. :)

## StatefulSets in Action prior to Istio 1.10

> In my GKE 1.19 cluster, I have Istio 1.9.5 installed. I enabled automatic sidecar injection in my default namespace, then I installed ZooKeeper using the [helm charts](https://artifacthub.io/packages/helm/bitnami/zookeeper) provided by Bitnami along with our `sleep` application:

Author: updated

$ helm install my-release bitnami/zookeeper --set replicaCount=3
$ kubectl apply -f {{< github_file >}}/samples/sleep/sleep.yaml
{{< /text >}}

After a few minutes, all pods come up nicely with sidecar proxies:

{{< text bash plain >}}
$ kubectl get pods,svc
NAME                             READY   STATUS    RESTARTS   AGE
my-release-zookeeper-0           2/2     Running   0          3h4m
my-release-zookeeper-1           2/2     Running   0          3h4m
my-release-zookeeper-2           2/2     Running   0          3h5m
pod/sleep-8f795f47d-qkgh4        2/2     Running   0          3h8m

NAME                            TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                            AGE
my-release-zookeeper            ClusterIP   10.100.1.113   <none>        2181/TCP,2888/TCP,3888/TCP         3h
my-release-zookeeper-headless   ClusterIP   None           <none>        2181/TCP,2888/TCP,3888/TCP         3h
service/sleep                           ClusterIP   10.100.9.26    <none>        80/TCP                             3h
{{< /text >}}

Are my ZooKeeper services working and the `READY` status correct? Let’s find out! ZooKeeper listens on 3 ports:

> *   port 2181 is the TCP port for clients to connect to the ZooKeeper service

Reviewer 2: ```suggestion
Are my ZooKeeper services working and is the status `READY`? Let’s find out! ZooKeeper listens on 3 ports:

*   port 2181 is the TCP port for clients to connect to the ZooKeeper service
```

doesn't render the list properly

Author: Yep good catch!

$ istioctl proxy-config cluster my-release-zookeeper-1 --port 3888 --direction inbound -o json
[
    {
        "name": "inbound|3888||",
        "type": "STATIC",
        "connectTimeout": "10s",
        "loadAssignment": {
            "clusterName": "inbound|3888||",
            "endpoints": [
                {
                    "lbEndpoints": [
                        {
                            "endpoint": {
                                "address": {
                                    "socketAddress": {
                                        "address": "127.0.0.1",
                                        "portValue": 3888
                                    }
                                }
                            }
                        }
                    ]
                }
            ]
        },
...
{{< /text >}}

What is interesting here is that the inbound on port 3888 has 127.0.0.1 as its endpoint. This means that Envoy proxy before Istio 1.10 redirects the inbound traffic to the `loopback` interface, as described in the [upcoming networking change in 1.10 blog](/blog/2021/upcoming-networking-changes/).

## StatefulSets in Action with Istio 1.10

> Install Istio 1.10 and configure the `default` namespace to enable Istio 1.10’s sidecar injection. Let’s rolling restart the ZooKeeper StatefulSet to update the pods to use Istio 1.10’s sidecar proxy:

Reviewer 2: you say Istio 1.10 a lot in this Istio 1.10 sentence, powered by Istio 1.10

Author: Open to suggestions on changing this :) I used 1.10 few times for clarity.

{{< text bash >}}
$ kubectl rollout restart statefulset my-release-zookeeper

> {{< /text >}}

Reviewer 2: Would it be more powerful if you installed again from the top?

(In fact, if that's the case, maybe you should lead with this section?)

Author: My thoughts is to meet users where they are today given these are very common issues with stateful sets and proved that we are better while explaining how the network behavior changed.

---

> title: "StatefulSets Made Easier With Istio 1.10"

Reviewer 2: Could you rename the file "statefulsets-made-easier" to match the CamelCase?

Author: done!

Author: Actually, @craigbox  this will cause lint error, so I'm rolling back the change.  This is probably why all the other blog folders use lower case only.

```
http://localhost:5153/blog/2021/statefulsets-made-easier/
45
  [404] http://localhost:5153/blog/2021/StatefulSets-made-easier/view-zookeeper-from-kiali.png
46
ERROR: Detected 1 broken links. Scanned 4000 links in 48.742 seconds.
```

Reviewer 2: Lower case please - "statefulsets-made-easier" should work.  (I was asking to remove the hyphen, not change the case 🙂)

description: Learn how to easily deploy StatefulSets with Istio 1.10.
publishdate: 2021-05-18
attribution: "Lin Sun (Solo.io), Christian Posta (Solo.io), John Howard (Google), Zhonghu Xu (Huawei)"
keywords: [statefulset,Istio,networking,localhost,loopback,eth0]
---

> Kubernetes [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) are commonly used to manage stateful applications. In addition to managing the deployment and scaling of a set of Pods, StatefulSets provide guarantees about the ordering and uniqueness of those Pods. Common applications used with StatefulSets include ZooKeeper, Cassandra, Elasticsearch, Redis and NiFi.

Reviewer 3: I will also add that Pod usually has backticks as well, at least when capitalized.

> The Istio community has been making gradual progress towards zero-configuration support for StatefulSets; from automatic mTLS, to eliminating the need to create DestinationRule or ServiceEntry resources, to the most recent [pod networking changes in Istio 1.10](/blog/2021/upcoming-networking-changes/).

Reviewer 3: Backtick on ServiceEntry

> What is unique about using a StatefulSet with a service mesh? The StatefulSet pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. It is common for pods within a StatefulSet listen on the pod IP only, instead of `0.0.0.0`, for communication within the StatefulSet pods.

Reviewer 3: **to** listen

Many [issues](https://github.com/istio/istio/issues/10659) have been reported from the Istio community around `StatefulSets`.

> Using ZooKeeper as an example, from the `zoo.cfg` file in our ZooKeeper pod, it is configured not to listen on all IPs for quorum communication within the `StatefulSet` pods.

Reviewer 3: nit: s/not to/to not/

Author: defer to @craigbox since he just changed this portion.

Reviewer 2: I agree with Eric, adding suggestion

> What is unique about using a `StatefulSet` with a service mesh? The `StatefulSet` pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. It is common for pods within a `StatefulSet` to listen on the pod IP only, instead of `0.0.0.0`, for communication within the `StatefulSet` pods.
> 
> Many [issues](https://github.com/istio/istio/issues/10659) have been reported from the Istio community around `StatefulSets`.
> 
> Using ZooKeeper as an example, from the `zoo.cfg` file in our ZooKeeper pod, it is configured not to listen on all IPs for quorum communication within the `StatefulSet` pods.
> 
> {{< text plain >}}
> quorumListenOnAllIPs=false
> {{< /text >}}

Reviewer 2: ```suggestion
What is unique about using a `StatefulSet` with a service mesh? The `StatefulSet` pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. The kind of apps that run in a `StatefulSet` are often those that need to communicate amongst their pods, and, as they come from a world of hard-coded IP addresses, may listen on the pod IP only, instead of `0.0.0.0`.

ZooKeeper, for example, is configured by default not to listen on all IPs for quorum communication:

{{< text plain >}}
quorumListenOnAllIPs=false
{{< /text >}}

Over the last few releases, the Istio community has [reported many issues](https://github.com/istio/istio/issues/10659) around support for applications running in `StatefulSets`.
```

Thoughts on that? @linsun @howardjohn

Author: Only thing I'm not sure is `and, as they come from a world of hard-coded IP addresses,` otherwise LGTM

@howardjohn ?

Reviewer: Sounds good enough to me, but I am not familiar with non-k8s world :slightly_smiling_face:

Author: Thanks Craig, we'll take your suggestions!

> ## `StatefulSets` in Action prior to Istio 1.10
> 
> In our GKE 1.19 cluster, we have Istio 1.9.5 installed. We enabled automatic sidecar injection in the `default` namespace, then we installed ZooKeeper using the [helm charts](https://artifacthub.io/packages/helm/bitnami/zookeeper) provided by Bitnami along with our `sleep` application:

Reviewer 2: ```suggestion
## `StatefulSets` in action, prior to Istio 1.10

In a GKE cluster running Kubernetes 1.19, we have Istio 1.9.5 installed. We enabled automatic sidecar injection in the `default` namespace, then we installed ZooKeeper using the [Helm charts provided by Bitnami](https://artifacthub.io/packages/helm/bitnami/zookeeper), along with our `sleep` sample application:
```

> In our GKE 1.19 cluster, we have Istio 1.9.5 installed. We enabled automatic sidecar injection in the `default` namespace, then we installed ZooKeeper using the [helm charts](https://artifacthub.io/packages/helm/bitnami/zookeeper) provided by Bitnami along with our `sleep` application:

Reviewer 3: nit: along with Istio's `sleep`

Author: I think okay to using `our` given the authors represent istio community here? @craigbox thoughts?

Reviewer 2: I'll clarify in a suggestion

{{< text bash yaml >}}
$ kubectl get pods,svc
NAME                             READY   STATUS    RESTARTS   AGE
my-release-zookeeper-0           2/2     Running   0          3h4m
my-release-zookeeper-1           2/2     Running   0          3h4m
my-release-zookeeper-2           2/2     Running   0          3h5m
pod/sleep-8f795f47d-qkgh4        2/2     Running   0          3h8m

NAME                            TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                            AGE
my-release-zookeeper            ClusterIP   10.100.1.113   <none>        2181/TCP,2888/TCP,3888/TCP         3h
my-release-zookeeper-headless   ClusterIP   None           <none>        2181/TCP,2888/TCP,3888/TCP         3h
service/sleep                   ClusterIP   10.100.9.26    <none>        80/TCP                             3h
{{< /text >}}

Are our ZooKeeper services working and is the status `Running`? Let’s find out! ZooKeeper listens on 3 ports:

* Port 2181 is the TCP port for clients to connect to the ZooKeeper service
* Port 2888 is the TCP port  for peers to connect to other peers
* Port 3888 is the dedicated TCP port for leader election

> By default, ZooKeeper installation configures port 2181 to listen on `0.0.0.0` but port 2888 and 3888 only listen on its pod IP. Let’s check out the network status on each of these ports from one of the ZooKeeper pods:

Reviewer 2: ```suggestion
By default, the ZooKeeper installation configures port 2181 to listen on `0.0.0.0` but ports 2888 and 3888 only listen on the pod IP. Let’s check out the network status on each of these ports from one of the ZooKeeper pods:
```

Client port found: 2181. Client address: localhost. Client SSL: false.
Error contacting service. It is probably not running.
{{< /text >}}

From the above output, you can see the ZooKeeper service is not functioning properly. Let us check the cluster configuration for one of the ZooKeeper pods:

{{< text bash yaml >}}
$ istioctl proxy-config cluster my-release-zookeeper-1 --port 3888 --direction inbound -o json
[
    {
        "name": "inbound|3888||",
        "type": "STATIC",
        "connectTimeout": "10s",
        "loadAssignment": {
            "clusterName": "inbound|3888||",
            "endpoints": [
                {
                    "lbEndpoints": [
                        {
                            "endpoint": {
                                "address": {
                                    "socketAddress": {
                                        "address": "127.0.0.1",
                                        "portValue": 3888
                                    }
                                }
                            }
                        }
                    ]
                }
            ]
        },
...
{{< /text >}}

> What is interesting here is that the inbound on port 3888 has `127.0.0.1` as its endpoint. This means that Envoy proxy before Istio 1.10 redirects the inbound traffic to the `loopback` interface, as described in the [upcoming networking change in 1.10 blog](/blog/2021/upcoming-networking-changes/).

Reviewer 2: ```suggestion
What is interesting here is that the inbound on port 3888 has `127.0.0.1` as its endpoint. This is because the Envoy proxy, in versions of Istio prior to 1.10, redirects the inbound traffic to the `loopback` interface, as described in the [our blog post about the change](/blog/2021/upcoming-networking-changes/).
```

> ## StatefulSets in Action with Istio 1.10

Reviewer 2: ```suggestion
## StatefulSets in action with Istio 1.10
```

Reviewer 2: You might need the backticks - I want it on the record that I think it's ugly as sin that we have to use them

> Install Istio 1.10 and configure the `default` namespace to enable Istio 1.10’s sidecar injection. Let’s rolling restart the ZooKeeper `StatefulSet` to update the pods to use Istio 1.10’s sidecar proxy:

Reviewer 2: ```suggestion
Now, we have upgraded our cluster to Istio 1.10 and configured the `default` namespace to enable 1.10 sidecar injection. Let’s rolling restart the ZooKeeper `StatefulSet` to update the pods to use the new version of the sidecar proxy:
```

(Perhaps we should say how we did this?  I'm assuming from the context this is an upgrade, not a second cluster)

{{< text bash >}}
$ kubectl rollout restart statefulset my-release-zookeeper
{{< /text >}}

> Once the ZooKeeper pods reach the running status, let’s check out the network connections for these 3 ports from any of the ZooKeeper pods:

Reviewer 2: ```suggestion
Once the ZooKeeper pods reach the `Running` status, we can check out the network connections for our 3 ports from any of the ZooKeeper pods:
```

tcp        0      0 10.96.8.10:44872        10.96.7.10:3888         ESTABLISHED
tcp        0      0 10.96.8.10:37878        10.96.6.11:3888         ESTABLISHED
tcp        0      0 10.96.8.10:44870        10.96.7.10:3888         ESTABLISHED
tcp        0      0 127.0.0.1:2181          127.0.0.1:54508         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54616         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54664         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54526         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54532         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54578         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54634         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54588         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54610         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54550         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54560         TIME_WAIT
tcp        0      0 127.0.0.1:2181          127.0.0.1:54644         TIME_WAIT
{{< /text >}}

> There are a few `ESTABLISHED` connections on both the 2888 and 3888 ports!  Next, let us check out the ZooKeeper server status.

Reviewer 2: ```suggestion
There are `ESTABLISHED` connections on both port 2888 and 3888!  Next, let us check out the ZooKeeper server status:
```

{{< text bash yaml >}}
$ kubectl exec my-release-zookeeper-1 -c zookeeper -- /opt/bitnami/zookeeper/bin/zkServer.sh status
/opt/bitnami/java/bin/java
ZooKeeper JMX enabled by default
Using config: /opt/bitnami/zookeeper/bin/../conf/zoo.cfg
Client port found: 2181. Client address: localhost. Client SSL: false.
Mode: follower
{{< /text >}}

> It is exciting that the ZooKeeper service appears to be running this time! Let’s connect to each of the ZooKeeper pods from the `sleep` pod and run the command below to discover the server status of each pod within the ZooKeeper `StatefulSet`. Note, there is no need to deploy ServiceEntry resources for any of the ZooKeeper pods and we can call these pods directly using their DNS names (e.g. `my-release-zookeeper-0.my-release-zookeeper-headless`) from the `sleep` pod.

Reviewer 2: ```suggestion
The ZooKeeper service is now running!

We can connect to each of the ZooKeeper pods from the `sleep` pod and run the below command to discover the server status of each pod within the `StatefulSet`. Note that there is no need to create ServiceEntry resources for any of the ZooKeeper pods and we can call these pods directly using their DNS names (e.g. `my-release-zookeeper-0.my-release-zookeeper-headless`) from the `sleep` pod.
```

{{< text bash yaml >}}
$ kubectl exec -it deploy/sleep -c sleep -- sh  -c 'for x in my-release-zookeeper-0.my-release-zookeeper-headless my-release-zookeeper-1.my-release-zookeeper-headless my-release-zookeeper-2.my-release-zookeeper-headless; do echo $x; echo srvr|nc $x 2181; echo; done'
my-release-zookeeper-0.my-release-zookeeper-headless
Zookeeper version: 3.7.0-e3704b390a6697bfdf4b0bef79e3da7a4f6bac4b, built on 2021-03-17 09:46 UTC
Latency min/avg/max: 1/7.5/20
Received: 3845
Sent: 3844
Connections: 1
Outstanding: 0
Zxid: 0x200000002
Mode: follower
Node count: 6

my-release-zookeeper-1.my-release-zookeeper-headless
Zookeeper version: 3.7.0-e3704b390a6697bfdf4b0bef79e3da7a4f6bac4b, built on 2021-03-17 09:46 UTC
Latency min/avg/max: 0/0.0/0
Received: 3856
Sent: 3855
Connections: 1
Outstanding: 0
Zxid: 0x200000002
Mode: follower
Node count: 6

my-release-zookeeper-2.my-release-zookeeper-headless
Zookeeper version: 3.7.0-e3704b390a6697bfdf4b0bef79e3da7a4f6bac4b, built on 2021-03-17 09:46 UTC
Latency min/avg/max: 0/0.0/0
Received: 3855
Sent: 3854
Connections: 1
Outstanding: 0
Zxid: 0x200000002
Mode: leader
Node count: 6
Proposal sizes last/min/max: 48/48/48
{{< /text >}}

> Now our ZooKeeper service is running, let’s use Istio to secure all communication to our ZooKeeper and ZooKeeper headless services. Apply mutual TLS to the `default` namespace:

Reviewer 2: ```suggestion
Now our ZooKeeper service is running, let’s use Istio to secure all communication to our regular and headless services. Apply mutual TLS to the `default` namespace:
```

{{< text bash >}}
$ kubectl apply -n default -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "default"
spec:
  mtls:
    mode: STRICT
EOF
{{< /text >}}

Continue sending some traffic from the `sleep` pod and bring up the Kiali dashboard to visualize the services in the `default` namespace:

> {{< image link="./view-zookeeper-from-kiali.png" caption="Visualize the ZooKeeper Services in Kiali" >}}

Reviewer 2: describe what you're seeing here?

"The padlock icons on the traffic flows indicate that the connections are secure."

Author: Thanks, added.

## Wrapping up

> With the new networking changes in Istio 1.10, a pod with a sidecar has consistent pod networking behavior as a pod without a sidecar in Kubernetes. This change enables Kubernetes `StatefulSets` or other types of applications to function properly in Istio as this blog has demonstrated. We believe this is a huge step towards Istio’s goal of providing transparent service mesh and zero-configuration Istio.

Reviewer 2: ```suggestion
With the new networking changes in Istio 1.10, a Kubernetes pod with a sidecar has the same networking behavior as a pod without a sidecar. This change enables stateful applications to function properly in Istio. as we have shown you in this post. We believe this is a huge step towards Istio’s goal of providing transparent service mesh and zero-configuration Istio.
```
